src/query.ts

```diff
@@ -0,0 +1,9 @@
+import QueryParamDescription from "./QueryParamDescription";
+
+type AllowedQueryParams = string | { [key: string]: AllowedQueryParams };
+
+export default function query<TReturnType extends AllowedQueryParams = string, TOptional extends boolean = false>(
+    optional: TOptional = false as TOptional,
```

Comment:
TBH, I'm not sure. I was thinking about that. But the problem is the way how you use it. You can have something like that:
```js
...segment`/reports`,
query: {
    clientId: query(true),
    teamId: query(true),
    productId: query(true),
    userId: query(true),
},
```
which is more consie than
```js
...segment`/reports`,
query: {
    clientId: query({ optional: true }),
    teamId: query({ optional: true }),
    productId: query({ optional: true }),
    userId: query({ optional: true }),
},
```
It can get noisy quickly, I think, and for now it's only option which is passed to query.

Maybe other thing which we can do is instead of `boolean` use string literal `"optional" | "required"`?

Reply:
That's a great idea! I would go for `"optional" | "required"` with the `"required"` option as a default, of course. It will be way more readable.